File: iOS/Free BeeTests/GameViewModelTest.swift
//
//  GameViewModelTest.swift
//  Free BeeTests
//
//  Created by G on 10/17/24.
//

import XCTest
@testable import Free_Bee

final class GameViewModelTest: XCTestCase {
   var persistenceController: PersistenceController!
   var game: Game!
   var viewModel: GameViewModel!
   
   override func setUpWithError() throws {
      // Put setup code here. This method is called before the invocation of each test method in the class.
      persistenceController = PersistenceController(inMemory: true)
      game = Game(context: persistenceController.container.viewContext)
      game.progress = GameProgress(context: game.managedObjectContext!)
      game.centerLetterCode = Int32("z".unicodeScalars.first!.value)
      game.otherLetters = "abcdef"
      game.geniusScore = 100
      game.maximumScore = 200
      game.date = Date(timeIntervalSinceReferenceDate: 0)
      game.allowedWords = ["abcd", "abcdefz"]
      viewModel = GameViewModel(game: game, objectContext: persistenceController.container.viewContext)
   }
   override func tearDownWithError() throws {
      // Put teardown code here. This method is called after the invocation of each test method in the class.
   }

   func test_enteredWordSummary() {
      XCTAssertEqual("No words yet", viewModel.enteredWordSummary)
      
      game.progress!.enteredWords = NSOrderedSet(array: ["abcd", "bcde"].map {
         EnteredWord(context: game.managedObjectContext!, string: $0)
      })
      XCTAssertEqual("Abcd\u{2003}Bcde", viewModel.enteredWordSummary)
   }
   
   func test_enterEnabled() {
      XCTAssertFalse(viewModel.enterEnabled, "enter disabled: current word empty")
      
      game.progress!.currentWord = "abc"
      XCTAssertFalse(viewModel.enterEnabled, "enter disabled: current word too short")

      game.progress!.currentWord = "abcde"
      XCTAssertFalse(viewModel.enterEnabled, "enter disabled: current word missing center letter")
      
      game.progress!.currentWord = "abcz"
      XCTAssertTrue(viewModel.enterEnabled, "enter enabled")
      
   }

   func test_append() {
      viewModel.append(letter: Character("a"))
      XCTAssertEqual("a", game.progress?.currentWord, "current word is \"a\"")
      
      viewModel.append(letter: "z")
      XCTAssertEqual("az", game.progress?.currentWord, "current word is \"az\"")
      
      game.progress?.currentWord = "abbcbcbcbcbcbcbbcbcbcbcb"
      viewModel.append(letter: "z")
      XCTAssertEqual("abbcbcbcbcbcbcbbcbcbcbcb", game.progress?.currentWord, "doesn't add letter, word max length")
   }
   
   func test_backspace() {
      viewModel.backspace()
      XCTAssertEqual("", game.progress!.currentWord)
      
      game.progress!.currentWord = "abcdef"
      viewModel.backspace()
      XCTAssertEqual("abcde", game.progress!.currentWord, "backspace")
   }
   
   func test_enter() {
      game.progress?.currentWord = "abcde"
      viewModel.enter()
      XCTAssertEqual(0, game.progress!.score, "score - word not accepted")
      XCTAssertEqual(0, game.progress!.enteredWords!.count, "enteredWords - word not accepted")
      XCTAssertEqual("", game.progress!.currentWord, "enter - currentWord cleared")
      
      game.progress?.currentWord = "abcdefz"
      viewModel.enter()
      XCTAssertEqual(14, game.progress!.score, "score - word accepted (pangram)")
      XCTAssertEqual(1, game.progress!.enteredWords!.count, "enteredWords - word accepted")
      XCTAssertEqual("", game.progress!.currentWord, "enter - currentWord cleared")

      game.progress?.currentWord = "abcd"
      viewModel.enter()
      XCTAssertEqual(15, game.progress!.score, "score - word accepted")
      XCTAssertEqual(2, game.progress!.enteredWords!.count, "enteredWords - word accepted")
      XCTAssertEqual("", game.progress!.currentWord, "enter - currentWord cleared")

      game.progress?.currentWord = "abcd"
      viewModel.enter()
      XCTAssertEqual(15, game.progress!.score, "score - repeat word not accepted")
      XCTAssertEqual(2, game.progress!.enteredWords!.count, "enteredWords - repeat word not accepted")
   }
   
   func test_onCurrentWordChanged() {
      let expectation = XCTestExpectation(description: "received onCurrentWordChanged")
      viewModel.onCurrentWordChanged = { word in
         XCTAssertEqual("abcd", word, "block param matches")
         expectation.fulfill()
      }
      
      game.progress!.currentWord = "abcd"
      wait(for: [expectation], timeout: 1)
   }
   
   func test_entryNotAcceptedEvent() {
      let expectation = XCTestExpectation(description: "received entryNotAcceptedEvent")
      let cancellable = viewModel.entryNotAcceptedEvent.sink {
         expectation.fulfill()
      }
      
      viewModel.enter()
      wait(for: [expectation], timeout: 1)
   }
}

File: KMM/composeApp/src/commonTest/kotlin/com/letstwinkle/freebee/unit/GameViewModelTest.kt
@file:Suppress("MemberVisibilityCanBePrivate")

package com.letstwinkle.freebee.unit

import com.letstwinkle.freebee.screens.game.GameViewModel
import com.letstwinkle.freebee.util.MockGame
import com.letstwinkle.freebee.util.TestRepository
import com.russhwolf.settings.MapSettings
import kotlinx.coroutines.*
import kotlinx.coroutines.channels.consumeEach
import kotlinx.coroutines.test.*
import kotlinx.datetime.LocalDate
import org.lighthousegames.logging.KmLogging
import org.lighthousegames.logging.LogLevel
import kotlin.test.*

private typealias TestGameViewModel = GameViewModel<Int, MockGame>

@OptIn(ExperimentalCoroutinesApi::class)
class GameViewModelTest {
   lateinit var viewModel: TestGameViewModel
   lateinit var repository: TestRepository
   
   val testDispatcher: TestDispatcher = StandardTestDispatcher()
   
   init {
      KmLogging.setLogLevel(LogLevel.Off)
   }
   
   @BeforeTest fun setUp() {
      Dispatchers.setMain(testDispatcher)
      repository = TestRepository()
      val gameID = runBlocking {
         val gameID = repository.createGame(
            date = LocalDate.fromEpochDays(0),
            allowedWords = setOf("zabc", "zdef", "zabcdef"),
            centerLetterCode = 'z'.code,
            otherLetters = "abcdef",
            geniusScore = 100,
            maximumScore = 200,
         )
         // 2nd game for referencing entered words
         repository.createGame(
            date = LocalDate.fromEpochDays(1),
            allowedWords = setOf("zggg", "zdef", "zhhh"),
            centerLetterCode = 'z'.code,
            otherLetters = "defghi",
            geniusScore = 100,
            maximumScore = 200,
         )
         val otherGame = repository.games.last()
         repository.addEnteredWord(otherGame, "zdef")
         gameID
      }
      viewModel = GameViewModel(
         repository = repository,
         gameID = gameID,
         multiplatformSettings = MapSettings(),
         backgroundContext = Dispatchers.Main
      )
   }
   
   @AfterTest fun tearDown() {
      Dispatchers.resetMain()
   }
   
   val gameWithWords: MockGame
      get() = viewModel.gameWithWords.value!!
   
   @Test fun testEnteredWordSummary() = runTest {
      assertEquals("No words yet", viewModel.enteredWordSummary, "Entered word summary when enteredWordCount = 0")
      
      gameWithWords.game.currentWord = "zabc"
      viewModel.enter()
      gameWithWords.game.currentWord = "zdef"
      viewModel.enter()
      
      assertEquals("Zdef\u2003Zabc", viewModel.enteredWordSummary, "Entered word summary when \"zabc\" and \"zdef\" are the entered words")
   }
   
   @Test fun testEnterEnabled() = runTest {
      val game = gameWithWords.game
      
      assertFalse(viewModel.enterEnabled, "enter disabled: currentWord is empty")
      
      game.currentWord = "abz"
      assertFalse(viewModel.enterEnabled, "enter disabled: currentWord too short")
      
      game.currentWord = "abcde"
      assertFalse(viewModel.enterEnabled, "enter disabled: currentWord missing center letter")
      
      game.currentWord = "abcdz"
      assertTrue(viewModel.enterEnabled, "enter enabled")
   }
   
   @Test fun testGameProgress() = runTest {
      assertEquals(0f, viewModel.gameProgress, absoluteTolerance = 0.001f, "game progress is 0")
      
      repository.addEnteredWord(gameWithWords, "zabc")
      assertEquals(0.333f, viewModel.gameProgress, absoluteTolerance = 0.001f, "game progress is 0.3…")
      
      repository.addEnteredWord(gameWithWords, "zdef")
      repository.addEnteredWord(gameWithWords, "zabcdef")
      assertEquals(1.0f, viewModel.gameProgress, absoluteTolerance = 0.001f, "game progress is 1.0")
   }
   
   @Test fun testAppend() = runTest {
      viewModel.append('a')
      assertEquals("a", gameWithWords.game.currentWord, "currentWord is \"a\"")
      
      viewModel.append('z')
      assertEquals("az", gameWithWords.game.currentWord, "currentWord is \"az\"")
      
      val tooLongWord = "a".repeat(50)
      gameWithWords.game.currentWord = tooLongWord
      viewModel.append('a')
      assertEquals(tooLongWord, gameWithWords.game.currentWord, "currentWord doesn't change - max length")
   }
   
   @Test fun testBackspace() = runTest {
      viewModel.backspace()
      assertEquals("", gameWithWords.game.currentWord, "backspace empty")
      
      gameWithWords.game.currentWord = "abcde"
      viewModel.backspace()
      assertEquals("abcd", gameWithWords.game.currentWord, "backspace not empty")
   }
   
   @Test fun testEnter() = runTest {
      backgroundScope.launch { 
         viewModel.entryNotAcceptedEvents.consumeEach {}
      }
      
      gameWithWords.game.currentWord = "abcde"
      viewModel.enter()
      assertEquals(0, gameWithWords.game.score, "testEnter: word not accepted: score")
      assertEquals(0, gameWithWords.enteredWords.size, "testEnter: word not accepted: enteredWords")
      assertEquals("", gameWithWords.game.currentWord, "testEnter: word not accepted: currentWord reset")
      
      gameWithWords.game.currentWord = "zabcdef"
      viewModel.enter()
      assertEquals(14, gameWithWords.game.score, "testEnter: word accepted (pangram): score")
      assertEquals(1, gameWithWords.enteredWords.size, "testEnter: word accepted (pangram): enteredWords")
      assertEquals("", gameWithWords.game.currentWord, "testEnter: word not accepted: currentWord reset")
      
      gameWithWords.game.currentWord = "zabcdef"
      viewModel.enter()
      assertEquals(14, gameWithWords.game.score, "testEnter: word not accepted, repeat: score")
      assertEquals(1, gameWithWords.enteredWords.size, "testEnter: word not accepted, repeat: enteredWords")
      assertEquals("", gameWithWords.game.currentWord, "testEnter: word not accepted, repeat: currentWord reset")
   }
   
   @Test fun testEntryNotAcceptedEvent() = runTest {
      launch {
         val event = viewModel.entryNotAcceptedEvents.receive()
         assertEquals("Entry isn't accepted", event, "currentWord invalid - receive event")
      }
      gameWithWords.game.currentWord = "zbcd"
      viewModel.enter()
      
      gameWithWords.game.currentWord = "zabc"
      viewModel.enter()
      assertTrue(viewModel.entryNotAcceptedEvents.isEmpty, "currentWord valid - no event")
   }
   
   @Test fun testWordHintsWhenNotEnoughWordsEntered() = runTest {
      advanceUntilIdle() // makes sure that success not coincidence
      assertTrue(viewModel.wordHints.value.isEmpty(), "wordHints - not enough words entered")
   }
   
   @Test fun testWordHintsWhenEnoughWordsEntered() = runTest {
      repository.addEnteredWord(repository.games.first(), "zabc")
      repository.addEnteredWord(repository.games.first(), "zabcdef")
      viewModel.initialize()
      advanceUntilIdle()
      
      val expectedHints = mapOf("zdef" to repository.games.last().date)
      assertEquals(expectedHints, viewModel.wordHints.value, "wordHints - enough words entered")
   }
}
